### ISSUE TITLE
Select operation does not work

### ISSUE THREAD

**Issue by mmnosek** [Starcounter/Starcounter.Linq#46]
Query like:

```c#
var existingItems = DbLinq.Objects<BasketItem>()
                   .Where(item => item.Basket == basket)
                   .Select(item => item.IntId)
                   .ToList();
```
Produces SQL query that returns whole `BasketItem` instead of it's `IntId`. When casting to the expected result (int), an exception is raised:

```
InvalidCastException: Unable to cast object of type 'Starcounter.Nova.Level1QPBridge.SqlResult`1[Microsoft.eShopWeb.ApplicationCore.Entities.BasketItem]' to type 'System.Collections.Generic.IEnumerable`1[System.Int32]'.
```

Issue related to Nova (`Linq v3.0.1`)

**Comment by un-tone** [comment]
Fixed in versions [1.1](https://github.com/Starcounter/Starcounter.Linq/releases/tag/1.1.0), [2.1](https://github.com/Starcounter/Starcounter.Linq/releases/tag/2.1.0), [3.1](https://github.com/Starcounter/Starcounter.Linq/releases/tag/3.1.0)

**Comment by un-tone** [comment]
cc @PatrykSzwer

**Comment by joozek78** [comment]
This code still fails on `2.1.0`:
```c#
using System.Linq;
using Starcounter;
using Starcounter.Linq;

namespace StarcounterApplication1
{
    [Database]
    public class Dog
    {
        public Person Owner { get; set; }
    }

    [Database]
    public class Person
    {
        public Location Location { get; set; }
    }

    [Database]
    public class Location
    {
    }

   
    class Program
    {
        static void Main()
        {
            DbLinq.Objects<Dog>()
                .Select(dog => dog.Owner)
                .Where(person => person.Location == null)
                .ToList();
        }
    }
}
```

Exception:
```
Starcounter.DbException: ScErrFailingEntrypoint (SCERR4259): The user code entrypoint method raised an exception. System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.Exception: ScErrQueryResultTypeMismatch (SCERR7031): Expected generic result type does not match with query result type. The result type StarcounterApplication1.Person cannot be assigned from the query result, which is of type StarcounterApplication1.Dog.
Version: 2.4.0.6619.
Help page: https://docs.starcounter.io/?q=SCERR7031.
   at Starcounter.Query.QueryPreparation.MatchEnumeratorResultAndExpectedType[T](IExecutionEnumerator execEnum)
   at Starcounter.Query.QueryPreparation.PrepareOrExecuteQuery[T](String appName, String query, Boolean slowSql, Object[] values)
   at Starcounter.Query.Sql.SqlEnumCache.CacheOrExecuteEnumerator[T](String appName, String query, Boolean slowSQL, Object[] values)
   at Starcounter.QueryProcessor.Db.FullQueryProcess[T](String query, Boolean slowSQL, String appName, Object[] values)
   at Starcounter.Db.FullQueryProcess[T](String query, Boolean slowSQL, String appName, Object[] values) in C:\	c_work\sc-19247\Level1\src\Starcounter\Mock\Db.cs:line 133
   at Starcounter.Db.SlowSQL[T](String query, Object[] values) in C:\	c_work\sc-19247\Level1\src\Starcounter\Mock\Db.cs:line 121
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Starcounter.Linq.QueryExecutor`1.Execute[TResult](String sql, Object[] variables, QueryResultMethod queryResultMethod)
   at Starcounter.Linq.QueryProvider.System.Linq.IQueryProvider.Execute[T](Expression expression)
   at Starcounter.Linq.Queryable`1.GetEnumerator()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at StarcounterApplication1.Program.Main() in C:\Users\joozek\dev\StarcounterApplication1\StarcounterApplication1\Program.cs:line 31
   at Starcounter.Hosting.Package.ExecuteEntryPoint(Application application, Assembly assembly) in C:\	c_work\sc-19247\Level1\src\Starcounter.Hosting\Hosting\Package.cs:line 550
Version: 2.4.0.6619.
Help page: https://docs.starcounter.io/?q=SCERR4259.
   at Starcounter.Hosting.Package.ExecuteEntryPoint(Application application, Assembly assembly) in C:\	c_work\sc-19247\Level1\src\Starcounter.Hosting\Hosting\Package.cs:line 584
   at Starcounter.Hosting.Package.ProcessWithinCurrentApplication(Application application, ApplicationDirectory applicationDir) in C:\	c_work\sc-19247\Level1\src\Starcounter.Hosting\Hosting\Package.cs:line 292
   at Starcounter.Hosting.Package.Process() in C:\	c_work\sc-19247\Level1\src\Starcounter.Hosting\Hosting\Package.cs:line 140
   at Starcounter.Hosting.Package.Process(IntPtr hPackage) in C:\	c_work\sc-19247\Level1\src\Starcounter.Hosting\Hosting\Package.cs:line 54
   at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in C:\	c_work\sc-19247\Level1\src\Starcounter.Hosting\StarcounterInternal.Hosting\Processor.cs:line 90
ErrorCode=4259
HResult=-2146233088
HelpLink=https://docs.starcounter.io/?q=SCERR4259
---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Starcounter.Linq.QueryExecutor`1.Execute[TResult](String sql, Object[] variables, QueryResultMethod queryResultMethod)
   at Starcounter.Linq.QueryProvider.System.Linq.IQueryProvider.Execute[T](Expression expression)
   at Starcounter.Linq.Queryable`1.GetEnumerator()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at StarcounterApplication1.Program.Main() in C:\Users\joozek\dev\StarcounterApplication1\StarcounterApplication1\Program.cs:line 31
   at Starcounter.Hosting.Package.ExecuteEntryPoint(Application application, Assembly assembly) in C:\	c_work\sc-19247\Level1\src\Starcounter.Hosting\Hosting\Package.cs:line 550
HResult=-2146232828
---> System.Exception: ScErrQueryResultTypeMismatch (SCERR7031): Expected generic result type does not match with query result type. The result type StarcounterApplication1.Person cannot be assigned from the query result, which is of type StarcounterApplication1.Dog.
Version: 2.4.0.6619.
Help page: https://docs.starcounter.io/?q=SCERR7031.
   at Starcounter.Query.QueryPreparation.MatchEnumeratorResultAndExpectedType[T](IExecutionEnumerator execEnum)
   at Starcounter.Query.QueryPreparation.PrepareOrExecuteQuery[T](String appName, String query, Boolean slowSql, Object[] values)
   at Starcounter.Query.Sql.SqlEnumCache.CacheOrExecuteEnumerator[T](String appName, String query, Boolean slowSQL, Object[] values)
   at Starcounter.QueryProcessor.Db.FullQueryProcess[T](String query, Boolean slowSQL, String appName, Object[] values)
   at Starcounter.Db.FullQueryProcess[T](String query, Boolean slowSQL, String appName, Object[] values) in C:\	c_work\sc-19247\Level1\src\Starcounter\Mock\Db.cs:line 133
   at Starcounter.Db.SlowSQL[T](String query, Object[] values) in C:\	c_work\sc-19247\Level1\src\Starcounter\Mock\Db.cs:line 121
HResult=-2146233088
HelpLink=https://docs.starcounter.io/?q=SCERR7031
```

**Comment by un-tone** [comment]
Calling `Where` after `Select` doesn't work correctly if in `Where` embedded property is referenced. Will be fixed. Thanks for reporting.

**Comment by joozek78** [comment]
Please also test for nested Selects, i.e. `DbLinq.Objects<Dog>.Select(dog => dog.Owner).Where(...).Select(person => person.Home)`. That was what I intended to test before I bumped into the issue I reported

**Comment by un-tone** [comment]
Fixed in [1.2](https://github.com/Starcounter/Starcounter.Linq/releases/tag/1.2.0), [2.2](https://github.com/Starcounter/Starcounter.Linq/releases/tag/2.2.0), [3.2](https://github.com/Starcounter/Starcounter.Linq/releases/tag/3.2.0). @joozek78 please check.
